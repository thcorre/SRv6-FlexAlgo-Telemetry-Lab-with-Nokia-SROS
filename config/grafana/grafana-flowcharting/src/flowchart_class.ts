import XGraph from 'graph_class';
import { StateHandler } from 'statesHandler';
import { State } from 'state_class';
import { Rule } from 'rule_class';

import { FlowchartHandler } from 'flowchartHandler';
import { Metric } from 'metric_class';
import { $GF } from 'globals_class';

/**
 * Flowchart handler
 *
 * @export
 * @class Flowchart
 */
export class Flowchart {
  data: gf.TFlowchartData;
  container: HTMLDivElement;
  xgraph: XGraph | undefined = undefined;
  stateHandler: StateHandler | undefined;
  // ctrl: any;
  templateSrv: any;
  states: Map<string, State> | undefined;

  constructor(name: string, container: HTMLDivElement, ctrl: any, data: gf.TFlowchartData) {
    this.data = data;
    this.data.name = name;
    this.container = container;
    this.templateSrv = ctrl.templateSrv;
  }

  /**
   * Import data object in current flowchart
   *
   * @param {Object} obj
   * @memberof Flowchart
   */
  import(obj: any): this {
    $GF.log.info(`flowchart[${this.data.name}].import()`);
    if (!!obj.download || this.data.download === false) {
      this.data.download = obj.download;
    }
    // 0.3.0
    if (!!obj.source) {
      this.data.type = obj.source.type;
      this.data.xml = obj.source.xml.value;
      this.data.url = obj.source.url.value;
    }
    // 0.3.0
    if (!!obj.options) {
      this.data.zoom = obj.options.zoom;
      this.data.center = obj.options.center;
      this.data.scale = obj.options.scale;
      this.data.lock = obj.options.lock;
      this.data.allowDrawio = false;
      this.data.tooltip = obj.options.tooltip;
      this.data.grid = obj.options.grid;
      this.data.bgColor = obj.options.bgColor;
    }
    if (!!obj.type) {
      this.data.type = obj.type;
    }
    if (!!obj.xml) {
      this.data.xml = obj.xml;
    }

    // 0.9.0
    if (!!obj.csv) {
      this.data.csv = obj.csv;
    }
    if (!!obj.url) {
      this.data.url = obj.url;
    }
    if (!!obj.zoom) {
      this.data.zoom = obj.zoom;
    }
    if (!!obj.center || obj.center === false) {
      this.data.center = obj.center;
    }
    if (!!obj.scale || obj.scale === false) {
      this.data.scale = obj.scale;
    }

    if (!!obj.lock || obj.lock === false) {
      this.data.lock = obj.lock;
    }
    if (!!obj.allowDrawio || obj.allowDrawio === false) {
      this.data.allowDrawio = obj.allowDrawio;
    }
    if (!!obj.enableAnim || obj.enableAnim === false) {
      this.data.enableAnim = obj.enableAnim;
    }
    if (!!obj.tooltip) {
      this.data.tooltip = obj.tooltip;
    }
    if (!!obj.grid || obj.grid === false) {
      this.data.grid = obj.grid;
    }
    if (!!obj.bgColor) {
      this.data.bgColor = obj.bgColor;
    }
    if (!!obj.editorUrl) {
      this.data.editorUrl = obj.editorUrl;
    }
    if (!!obj.editorTheme) {
      this.data.editorTheme = obj.editorTheme;
    }
    this.init();
    return this;
  }

  /**
   * Reset/empty/destroy flowchart
   *
   * @returns {this}
   * @memberof Flowchart
   */
  clear(): this {
    if (this.xgraph) {
      this.xgraph.destroyGraph();
      this.xgraph = undefined;
      this.container.remove();
    }
    if (this.stateHandler) {
      this.stateHandler.clear();
      this.stateHandler = undefined;
    }
    return this;
  }

  /**
   * Return the default XML when new
   *
   * @static
   * @returns {gf.TFlowchartData}
   * @memberof Flowchart
   */
  static getDefaultData(): gf.TFlowchartData {
    return {
      name: 'name',
      xml: FlowchartHandler.getDefaultDioGraph(),
      csv: FlowchartHandler.getDefaultCsvGraph(),
      download: false,
      type: 'xml',
      url: 'http://<YourUrl>/<Your XML/drawio file/api>',
      zoom: '100%',
      center: true,
      scale: true,
      lock: true,
      allowDrawio: false,
      enableAnim: true,
      tooltip: true,
      grid: false,
      bgColor: null,
      editorUrl: 'https://www.draw.io',
      editorTheme: 'dark',
    };
  }

  /**
   * Return data without functions to save json in grafana
   *
   * @returns {Object} Data object
   * @memberof Flowchart
   */
  getData(): gf.TFlowchartData {
    return this.data;
  }

  /**
   * Update states of flowchart/graph
   *
   * @param {*} rules
   * @returns {this}
   * @memberof Flowchart
   */
  updateStates(rules: Rule[]): this {
    const trc = $GF.trace.before(this.constructor.name + '.' + 'updateStates()');
    rules.forEach(rule => {
      if (this.stateHandler !== undefined) {
        rule.states = this.stateHandler.getStatesForRule(rule);
        if (rule.states) {
          rule.states.forEach((state: any) => {
            state.unsetState();
          });
        } else {
          $GF.log.warn('States not defined for this rule');
        }
      } else {
        $GF.log.error('updateStates => this.stateHandler undefined');
      }
    });
    trc.after();
    return this;
  }

  /**
   * Initialisation of flowchart class
   *
   * @return {this}
   * @memberof Flowchart
   */
  init(): this {
    try {
      const content = this.getContent();
      if (this.xgraph === undefined) {
        this.xgraph = new XGraph(this.container, this.data.type, content);
      }
      if (content !== undefined && content !== null) {
        // if (this.data.download) {
        //   this.xgraph.setXmlGraph(this.getContent());
        // }
        if (this.data.allowDrawio) {
          this.xgraph.allowDrawio(true);
        } else {
          this.xgraph.allowDrawio(false);
        }
        if (this.data.enableAnim) {
          this.xgraph.enableAnim(true);
        } else {
          this.xgraph.enableAnim(false);
        }
        this.setOptions();
        this.xgraph.drawGraph();
        if (this.data.tooltip) {
          this.xgraph.tooltipGraph(true);
        }
        if (this.data.scale) {
          this.xgraph.scaleGraph(true);
        } else {
          this.xgraph.zoomGraph(this.data.zoom);
        }
        if (this.data.center) {
          this.xgraph.centerGraph(true);
        }
        if (this.data.lock) {
          this.xgraph.lockGraph(true);
        }
        this.stateHandler = new StateHandler(this.xgraph);
      } else {
        $GF.log.error('Source content empty Graph not defined');
      }
    } catch (error) {
      $GF.log.error('Unable to initialize graph', error);
    }
    return this;
  }

  /**
   * Get states handler
   *
   * @returns
   * @memberof Flowchart
   */
  getStateHandler() {
    return this.stateHandler;
  }

  /**
   * Get XGraph
   *
   * @returns
   * @memberof Flowchart
   */
  getXGraph() {
    return this.xgraph;
  }

  /**
   * Init states with rules and metrics
   *
   * @param {Rule[]} rules
   * @param {Metric[]} metrics
   * @memberof Flowchart
   */
  setStates(rules: Rule[], metrics: Metric[]): this {
    $GF.log.info(`flowchart[${this.data.name}].setStates()`);
    if (rules === undefined) {
      $GF.log.warn("Rules shoudn't be null");
    }
    if (metrics === undefined) {
      $GF.log.warn("Metrics shoudn't be null");
    }
    if (this.stateHandler) {
      this.stateHandler.setStates(rules, metrics);
    }
    return this;
  }

  /**
   * Init options of graph
   *
   * @memberof Flowchart
   */
  setOptions(): this {
    const trc = $GF.trace.before(this.constructor.name + '.' + 'setOptions()');
    this.setScale(this.data.scale);
    this.setCenter(this.data.center);
    this.setGrid(this.data.grid);
    this.setTooltip(this.data.tooltip);
    this.setLock(this.data.lock);
    this.setZoom(this.data.zoom);
    this.setBgColor(this.data.bgColor);
    trc.after();
    return this;
  }

  /**
   * Apply new states (colors,text ...)
   *
   * @memberof Flowchart
   */
  applyStates(): this {
    $GF.log.info(`flowchart[${this.data.name}].applyStates()`);
    if (this.stateHandler) {
      this.stateHandler.applyStates();
    }
    return this;
  }

  /**
   * Apply options
   *
   * @memberof Flowchart
   */
  applyOptions() {
    const trc = $GF.trace.before(this.constructor.name + '.' + 'applyOptions()');
    if (this.xgraph) {
      this.xgraph.applyGraph();
    }
    trc.after();
  }

  /**
   * Refresh graph
   *
   * @memberof Flowchart
   */
  refresh() {
    if (this.xgraph) {
      this.xgraph.refresh();
    }
  }

  /**
   * Reset and redraw graph when source changed
   *
   * @param {*} xmlGraph
   * @memberof Flowchart
   */
  redraw(content?: string) {
    if (content !== undefined) {
      this.setContent(content);
    }
    if (this.xgraph !== undefined) {
      this.xgraph.setContent(this.getContent());
    }
    this.applyOptions();
  }

  /**
   * Reload source of graph
   *
   * @memberof Flowchart
   */
  reload() {
    if (this.xgraph !== undefined && this.xgraph !== null) {
      this.xgraph.destroyGraph();
      this.xgraph = undefined;
      this.init();
    } else {
      this.init();
    }
  }

  /**
   * Set paramater lock
   *
   * @param {boolean} bool
   * @returns {this}
   * @memberof Flowchart
   */
  setLock(bool: boolean): this {
    this.data.lock = bool;
    if (this.xgraph) {
      this.xgraph.lock = bool;
    }
    return this;
  }

  /**
   * Lock graph
   *
   * @param {boolean} bool
   * @returns {this}
   * @memberof Flowchart
   */
  applyLock(bool: boolean): this {
    if (bool !== undefined) {
      this.data.lock = bool;
    }
    if (this.xgraph) {
      this.xgraph.lockGraph(this.data.lock);
    }
    return this;
  }

  /**
   * Set enable tooltip
   *
   * @param {boolean} bool
   * @returns {this}
   * @memberof Flowchart
   */
  setTooltip(bool: boolean): this {
    this.data.tooltip = bool;
    if (this.xgraph) {
      this.xgraph.tooltip = bool;
    }
    return this;
  }

  /**
   * Enable tooltip
   *
   * @param {boolean} bool
   * @returns {this}
   * @memberof Flowchart
   */
  applyTooltip(bool: boolean): this {
    if (bool !== undefined) {
      this.data.tooltip = bool;
    }
    if (this.xgraph) {
      this.xgraph.tooltipGraph(this.data.tooltip);
    }
    return this;
  }

  /**
   * Set scale parameter
   *
   * @param {boolean} bool
   * @returns {this}
   * @memberof Flowchart
   */
  setScale(bool: boolean): this {
    this.data.scale = bool;
    if (this.xgraph) {
      this.xgraph.scale = bool;
    }
    return this;
  }

  /**
   * Set BgColor
   *
   * @param {(string | null)} bgColor
   * @returns {this}
   * @memberof Flowchart
   */
  setBgColor(bgColor: string | null): this {
    this.data.bgColor = bgColor;
    if (this.xgraph) {
      this.xgraph.bgColor = bgColor;
    }
    return this;
  }

  /**
   * Apply Background color
   *
   * @param {string} bgColor
   * @returns {this}
   * @memberof Flowchart
   */
  ApplyBgColor(bgColor: string): this {
    this.data.bgColor = bgColor;
    if (bgColor) {
      if (this.xgraph) {
        this.xgraph.bgGraph(bgColor);
      }
    }
    return this;
  }

  /**
   * Apply scale parameter
   *
   * @param {boolean} bool
   * @returns {this}
   * @memberof Flowchart
   */
  applyScale(bool: boolean): this {
    if (bool !== undefined) {
      this.data.scale = bool;
    }
    if (this.xgraph) {
      this.xgraph.scaleGraph(this.data.scale);
    }
    return this;
  }

  /**
   * set center parameter
   *
   * @param {boolean} bool
   * @returns
   * @memberof Flowchart
   */
  setCenter(bool: boolean) {
    this.data.center = bool;
    if (this.xgraph) {
      this.xgraph.center = bool;
    }
    return this;
  }

  /**
   * Get names array of names according id or value
   *
   * @param {gf.TPropertieKey} prop
   * @returns {string[]}
   * @memberof Flowchart
   */
  getNamesByProp(prop: gf.TPropertieKey): string[] {
    if (this.xgraph) {
      return this.xgraph.getOrignalCells(prop);
    }
    return [];
  }

  /**
   * get XML def with var replaced
   *
   * @param {boolean} replaceVarBool
   * @returns {string}
   * @memberof Flowchart
   */
  getXml(replaceVarBool: boolean = true): string {
    if (!replaceVarBool) {
      return this.data.xml;
    }
    return this.templateSrv.replaceWithText(this.data.xml);
  }

  /**
   * get CSV def with var replaced
   *
   * @param {boolean} replaceVarBool
   * @returns {string}
   * @memberof Flowchart
   */
  getCsv(replaceVarBool: boolean = true): string {
    if (!replaceVarBool) {
      return this.data.csv;
    }
    return this.templateSrv.replaceWithText(this.data.csv);
  }

  /**
   * Get data source according type
   *
   * @param {boolean} [replaceVarBool=true]
   * @returns
   * @memberof Flowchart
   */
  getSource(replaceVarBool: boolean = true) {
    if (this.data.type === 'xml') {
      return this.getXml(replaceVarBool);
    }
    if (this.data.type === 'csv') {
      return this.getCsv(replaceVarBool);
    }
    return '';
  }

  /**
   * Get Url editor
   *
   * @returns {string}
   * @memberof Flowchart
   */
  getUrlEditor(): string {
    return this.data.editorUrl;
  }

  getThemeEditor(): string {
    return this.data.editorTheme;
  }

  /**
   * Get Source of graph (csv|xml) or get content from url
   *
   * @param {boolean} replaceVarBool
   * @returns
   * @memberof Flowchart
   */
  getContent(replaceVarBool: boolean = true): string {
    const trc = $GF.trace.before(this.constructor.name + '.' + 'getContent()');
    let content: string | null = '';
    if (this.data.download) {
      const url = this.templateSrv.replaceWithText(this.data.url);
      content = this.loadContent(url);
      if (content !== null) {
        if (replaceVarBool) {
          content = this.templateSrv.replaceWithText(content);
        }
      }
    } else {
      content = this.getSource(replaceVarBool);
    }
    trc.after();
    return content === null ? '' : content;
  }

  /**
   * Set the data source
   *
   * @param {string} content
   * @returns {this}
   * @memberof Flowchart
   */
  setContent(content: string): this {
    if (this.data.type === 'xml') {
      this.data.xml = content;
    }
    if (this.data.type === 'csv') {
      this.data.csv = content;
    }
    return this;
  }

  /**
   * Load source from url
   *
   * @param {*} url
   * @returns
   * @memberof Flowchart
   */
  loadContent(url: string): string | null {
    return $GF.utils.$loadFile(url);
  }

  renameId(oldId: string, newId: string): this {
    if (this.xgraph) {
      this.xgraph.renameId(oldId, newId);
    }
    return this;
  }

  /**
   * Apply xml to graph
   *
   * @returns {this}
   * @memberof Flowchart
   */
  applyModel(): this {
    if (this.xgraph) {
      this.data.xml = this.xgraph.getXmlModel();
    }
    this.redraw();
    return this;
  }

  center(bool: boolean): this {
    if (bool !== undefined) {
      this.data.center = bool;
    }
    if (this.xgraph) {
      this.xgraph.centerGraph(this.data.center);
    }
    return this;
  }

  setZoom(percent: string): this {
    this.data.zoom = percent;
    if (this.xgraph) {
      this.xgraph.zoomPercent = percent;
    }
    return this;
  }

  zoom(percent: string): this {
    if (percent !== undefined) {
      this.data.zoom = percent;
    }
    if (this.xgraph) {
      this.xgraph.zoomGraph(this.data.zoom);
    }
    return this;
  }

  setGrid(bool: boolean): this {
    this.data.grid = bool;
    if (this.xgraph) {
      this.xgraph.grid = bool;
    }
    return this;
  }

  grid(bool: boolean): this {
    if (bool !== undefined) {
      this.data.grid = bool;
    }
    if (this.xgraph) {
      this.xgraph.gridGraph(this.data.grid);
    }
    return this;
  }

  setXml(xml: string): this {
    this.data.xml = xml;
    return this;
  }

  setCsv(csv: string): this {
    this.data.csv = csv;
    return this;
  }

  minify() {
    this.data.xml = $GF.utils.minify(this.data.xml);
  }

  prettify() {
    this.data.xml = $GF.utils.prettify(this.data.xml);
  }

  decode() {
    if ($GF.utils.isencoded(this.data.xml)) {
      this.data.xml = $GF.utils.decode(this.data.xml, true, true, true);
      // this.data.xml = XGraph.decompress(this.data.xml);
    }
  }

  encode() {
    if (!$GF.utils.isencoded(this.data.xml)) {
      this.data.xml = $GF.utils.encode(this.data.xml, true, true, true);
      // this.data.xml = XGraph.compress(this.data.xml);
    }
  }

  getContainer(): HTMLDivElement {
    return this.container;
  }

  setMap(onMappingObj: gf.TIOnMappingObj) {
    const container = this.getContainer();
    if (this.xgraph) {
      this.xgraph.setMap(onMappingObj);
    }
    container.scrollIntoView();
    container.focus();
  }

  unsetMap() {
    if (this.xgraph) {
      this.xgraph.unsetMap();
    }
  }
}
