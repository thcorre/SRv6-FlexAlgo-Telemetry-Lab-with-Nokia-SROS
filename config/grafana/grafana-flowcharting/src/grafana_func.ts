// import { getValueFormat } from '@grafana/data';
import kbn from 'grafana/app/core/utils/kbn';
import TimeSeries from 'grafana/app/core/time_series2';
import _ from 'lodash';
import { loadPluginCss, MetricsPanelCtrl } from 'grafana/app/plugins/sdk';
import { dateTime, PanelEvents } from '@grafana/data';
import appEvents from 'grafana/app/core/app_events';

var grafana = {
  // formatValue(value, unit, decimals) {
  //   const fmt = getValueFormat(unit);
  //   const result = fmt(value, decimals).toString();
  //   return result.toString();
  // }
  formatValue(value, unit, decimals) {
    return kbn.valueFormats[unit](value, decimals, null).toString();
  },

  getUnitFormats() {
    return kbn.getUnitFormats();
  },

  loadCss() {
    loadPluginCss({
      dark: 'plugins/agenty-flowcharting-panel/static/css/flowchart.dark.css',
      light: 'plugins/agenty-flowcharting-panel/static/css/flowchart.light.css',
    });
  },

  getTimeSeries(seriesData) {
    return new TimeSeries({
      datapoints: seriesData.datapoints || [],
      alias: seriesData.target,
      unit: seriesData.unit,
    });
  },

  getFormatedDate(value, format) {
    return dateTime(value).format(format);
  },
  appEvents: appEvents,
  MetricsPanelCtrl: MetricsPanelCtrl,
  PanelEvents: PanelEvents,
};

export default grafana;
