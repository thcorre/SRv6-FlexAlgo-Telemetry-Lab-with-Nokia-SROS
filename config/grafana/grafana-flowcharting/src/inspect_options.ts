import { FlowchartHandler } from './flowchartHandler';
import { State } from 'state_class';
import { $GF } from 'globals_class';
import { StateHandler } from 'statesHandler';
import _ from 'lodash';
// import { MetricHandler } from './metricHandler';

declare interface TColumn {
  id: string;
  desc: string;
  sort: 'asc' | 'desc';
  select: boolean;
}

declare interface TTable {
  data: State[];
  columns: TColumn[];
}

export class InspectOptionsCtrl {
  enable = false; // enable inspector or not
  ctrl: any; //TODO: define type
  flowchartHandler: FlowchartHandler;
  stateHandler: StateHandler | undefined;
  state: TTable;
  panel: any;
  traceEnable: boolean = $GF.trace.isEnabled();

  /** @ngInject */
  constructor($scope: gf.TInspectOptionsScope) {
    $scope.editor = this;
    $scope.$GF = $GF.me();

    this.state = {
      data: this.getStates(),
      columns: [
        {
          id: 'cellId',
          desc: 'Shape ID',
          sort: 'asc',
          select: false,
        },
        {
          id: 'label',
          desc: 'Label',
          sort: 'asc',
          select: false,
        },
        {
          id: 'cellId',
          desc: 'Shape ID',
          sort: 'asc',
          select: false,
        },
        {
          id: 'font',
          desc: 'Font color',
          sort: 'asc',
          select: false,
        },
        {
          id: 'fill',
          desc: 'Fill color',
          sort: 'asc',
          select: false,
        },
        {
          id: 'stroke',
          desc: 'Stroke color',
          sort: 'asc',
          select: false,
        },
        {
          id: 'tags',
          desc: 'Tags Mapping',
          sort: 'asc',
          select: false,
        },
      ],
    };
    this.ctrl = $scope.ctrl;
    this.panel = this.ctrl.panel;
    this.flowchartHandler = this.ctrl.flowchartHandler;
    this.stateHandler = this.flowchartHandler.getFlowchart().getStateHandler();
  }

  render() {
    this.panel.render();
  }

  // onDebug() {
  //   GFP.log.logLevel = this.logLevel;
  //   GFP.log.logDisplay = this.logDisplay;
  // }

  onChangeId(state: State) {
    if (state.newcellId !== undefined && state.cellId !== state.newcellId) {
      state.edited = true;
      const sh = this.flowchartHandler.getFlowchart().getStateHandler();
      if (sh !== undefined) {
        sh.edited = true;
      }
      if (state.previousId === undefined) {
        state.previousId = state.cellId;
      }
      state.cellId = state.newcellId;
      state.edited = true;
    }
    state.edit = false;
  }

  onEdit(state: State) {
    state.edit = true;
    state.newcellId = state.cellId;
    // let stateHandler = this.flowchartHandler.getFlowchart().getStateHandler();
    // stateHandler.edited = true;
    const elt = document.getElementById(state.cellId);
    setTimeout(() => {
      if (elt) {
        elt.focus();
      }
    }, 100);
  }

  reset() {
    const flowchart = this.flowchartHandler.getFlowchart();
    const sh = flowchart.getStateHandler();
    if (sh !== undefined) {
      const states = sh.getStates();
      states.forEach(state => {
        if (state.edited && state.previousId) {
          state.cellId = state.previousId;
          state.edited = false;
        }
      });
      sh.edited = false;
    }
    flowchart.applyModel();
    this.flowchartHandler.draw();
    this.flowchartHandler.refresh();
    // this.$scope.$apply();
  }

  apply() {
    const flowchart = this.flowchartHandler.getFlowchart();
    const sh = flowchart.getStateHandler();
    if (sh !== undefined) {
      const states = sh.getStates();
      states.forEach(state => {
        if (state.edited && state.previousId) {
          flowchart.renameId(state.previousId, state.cellId);
          state.edited = false;
        }
      });
      sh.edited = false;
    }
    flowchart.applyModel();
  }

  selectCell(state: State) {
    state.highlightCell();
  }

  unselectCell(state: State) {
    state.unhighlightCell();
  }

  getStates(): State[] {
    if (this.stateHandler) {
      const states = this.stateHandler.getStatesForInspect();
      return _.orderBy(states, ['cellId', 'globalLevel'], ['asc']);
    }
    return [];
  }

  getStateValue(state: State, col: string): string | null {
    switch (col) {
      case 'cellId':
        return state.cellId;
        break;
      case 'level':
        return state.getTextLevel();
        break;
      case 'label':
        return state.getCellProp('value');
        break;

      default:
        return null;
        break;
    }
  }

  tracePerf() {
    if (this.traceEnable) {
      $GF.trace.enable();
    } else {
      $GF.trace.disable();
    }
  }

  anonymize() {
    const fc = this.flowchartHandler.getFlowchart();
    const xg = fc.getXGraph();
    if (xg) {
      xg.anonymize();
    }
  }
}

/** @ngInject */
export function inspectOptionsTab($q, uiSegmentSrv) {
  'use strict';
  return {
    restrict: 'E',
    scope: true,
    templateUrl: `${$GF.plugin.getPartialPath()}inspect/index.html`,
    controller: InspectOptionsCtrl,
  };
}
