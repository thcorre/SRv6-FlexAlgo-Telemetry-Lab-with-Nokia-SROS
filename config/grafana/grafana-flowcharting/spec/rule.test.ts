import { Rule } from '../src/rule_class';

describe('Rule', () => {
  const data = Rule.getDefaultData();

  describe('New', () => {
    const rule = new Rule('/.*/', data);
    test('Should be equals', () => {
      expect(rule.getData()).toMatchSnapshot();
    });
  });

  describe('Level', () => {
    const rule = new Rule('/.*/', data);
    test('In order', () => {
      rule.data.thresholds = [50, 80];
      expect(rule.data.invert).toBeFalsy();
      expect(rule.getThresholdLevel(25)).toBe(2);
      expect(rule.getThresholdLevel(60)).toBe(1);
      expect(rule.getThresholdLevel(90)).toBe(0);
    });

    test('Invert', () => {
      rule.invertThesholds();
      expect(rule.data.invert).toBeTruthy();
      expect(rule.getThresholdLevel(25)).toBe(0);
      expect(rule.getThresholdLevel(60)).toBe(1);
      expect(rule.getThresholdLevel(90)).toBe(2);
    });
  });

  describe('Colors', () => {
    const rule = new Rule('/.*/', data);
    rule.data.thresholds = [50, 80];
    test('Each colors', () => {
      expect(rule.getColorForValue(25)).toBe('rgba(50, 172, 45, 0.97)');
      expect(rule.getColorForValue(60)).toBe('rgba(237, 129, 40, 0.89)');
      expect(rule.getColorForValue(90)).toBe('rgba(245, 54, 54, 0.9)');
    });
    test('Each colors', () => {
      rule.invertThesholds();
      expect(rule.getColorForValue(25)).toBe('rgba(245, 54, 54, 0.9)');
      expect(rule.getColorForValue(60)).toBe('rgba(237, 129, 40, 0.89)');
      expect(rule.getColorForValue(90)).toBe('rgba(50, 172, 45, 0.97)');
    });
  });

  describe('Shape', () => {
    const pattern = '/.*Toto.*/';

    describe('Create', () => {
      test.skip('addshape & remove', () => {
        const rule = new Rule('/.*/', data);
        rule.addShapeMap(pattern);
        expect(rule.getShapeMaps().length).toBe(1);
        expect(rule.data.shapeData.length).toBe(1);
        rule.removeShapeMap(0);
        expect(rule.getShapeMaps().length).toBe(0);
        expect(rule.data.shapeData.length).toBe(0);
      });
    });

    describe('Match', () => {
      test('Match on one', () => {
        const rule = new Rule('/.*/', data);
        rule.addShapeMap(pattern);
        const sm = rule.getShapeMap(0);
        sm.match('Toto');
        expect(sm.match('Toto')).toBeTruthy();
      });

      test('Match on all', () => {
        const rule = new Rule('/.*/', data);
        rule.addShapeMap(pattern);
        expect(rule.matchShape('Toto')).toBeTruthy();
        expect(rule.matchShape('Tata')).toBeFalsy();
      });
    });

    describe('Colorize', () => {
      const rule = new Rule('/.*/', data);
      rule.data.thresholds = [50, 80];
      const shape = rule.addShapeMap(pattern);

      test('toColorize always ERR should be true', () => {
        shape.data.colorOn = 'a';
        expect(shape.toColorize(1)).toBeTruthy();
      });
      test('toColorize Warn/err ERR should be true', () => {
        shape.data.colorOn = 'wc';
        expect(shape.toColorize(0)).toBeFalsy();
      });
      test('toColorize Warn/err WARN should be true', () => {
        shape.data.colorOn = 'wc';
        expect(shape.toColorize(1)).toBeTruthy();
      });
      test('toColorize Warn/err OK should be false', () => {
        shape.data.colorOn = 'wc';
        expect(shape.toColorize(2)).toBeTruthy();
      });
    });
  });

  describe('Text', () => {
    const pattern = '/.*Toto.*/';

    test('addshape & remove', () => {
      const rule = new Rule('/.*/', data);
      rule.addTextMap(pattern);
      expect(rule.getTextMaps().length).toBe(1);
      rule.removeTextMap(0);
      expect(rule.getTextMaps().length).toBe(0);
    });

    test('Match on one', () => {
      const rule = new Rule('/.*/', data);
      rule.addTextMap(pattern);
      const sm = rule.getTextMap(0);
      expect(sm.match('Toto')).toBeTruthy();
    });

    test('Match on all', () => {
      const rule = new Rule('/.*/', data);
      rule.addTextMap(pattern);
      expect(rule.matchText('Toto')).toBeTruthy();
      expect(rule.matchText('Tata')).toBeFalsy();
    });
  });

  describe('Link', () => {
    const pattern = '/.*Toto.*/';
    test('addshape & remove', () => {
      const rule = new Rule('/.*/', data);
      rule.addLinkMap(pattern);
      expect(rule.getLinkMaps().length).toBe(1);
      expect(rule.data.linkData.length).toBe(1);
      rule.removeLinkMap(0);
      expect(rule.getLinkMaps().length).toBe(0);
    });

    test('Match on one', () => {
      const rule = new Rule('/.*/', data);
      rule.addLinkMap(pattern);
      const sm = rule.getLinkMap(0);
      expect(sm.match('Toto')).toBeTruthy();
    });

    test('Match on all', () => {
      const rule = new Rule('/.*/', data);
      rule.addLinkMap(pattern);
      expect(rule.matchLink('Toto')).toBeTruthy();
      expect(rule.matchLink('Tata')).toBeFalsy();
    });
  });

  describe('ValueMap', () => {
    test('addshape & remove', () => {
      const rule = new Rule('/.*/', data);
      rule.data.mappingType = 1;
      rule.addValueMap(1, 'This 1');
      expect(rule.getValueMaps().length).toBe(1);
      expect(rule.data.valueData.length).toBe(1);
      rule.addValueMap(2, 'This 2');
      expect(rule.getValueMaps().length).toBe(2);
      expect(rule.data.valueData.length).toBe(2);
    });

    test('Mapping values', () => {
      const rule = new Rule('/.*/', data);
      rule.data.mappingType = 1;
      rule.data.type = 'string';
      rule.addValueMap(1, 'This 1');
      rule.addValueMap(2, 'This 2');
      expect(rule.getFormattedValue(1)).toBe('This 1');
      expect(rule.getFormattedValue(2)).toBe('This 2');
      expect(rule.getFormattedValue(3)).toBe('3');
    });
  });

  describe('RangeMap', () => {
    test('addshape & remove', () => {
      const rule = new Rule('/.*/', data);
      rule.data.mappingType = 2;
      rule.data.type = 'string';
      rule.addRangeMap(0, 5, 'Between 0 and 5');
      expect(rule.getRangeMaps().length).toBe(1);
      expect(rule.data.rangeData.length).toBe(1);
      rule.addRangeMap(6, 10, 'Between 6 and 10');
      expect(rule.getRangeMaps().length).toBe(2);
      expect(rule.data.rangeData.length).toBe(2);
    });

    test('Mapping values', () => {
      const rule = new Rule('/.*/', data);
      rule.data.mappingType = 2;
      rule.data.type = 'string';
      rule.addRangeMap(0, 5, 'Between 0 and 5');
      rule.addRangeMap(6, 10, 'Between 6 and 10');
      expect(rule.getFormattedValue(3)).toBe('Between 0 and 5');
      expect(rule.getFormattedValue(7)).toBe('Between 6 and 10');
      expect(rule.getFormattedValue(12)).toBe('12');
    });

    //TODO: Fix It
    test('Mapping values upper and lower', () => {
      const rule = new Rule('/.*/', data);
      rule.data.mappingType = 2;
      rule.data.type = 'string';
      rule.addRangeMap(undefined, 33, 'Lower than 33');
      rule.addRangeMap(33, 66, 'Between 33 and 66');
      rule.addRangeMap(66, undefined, 'Upper than 66');
      expect(rule.getFormattedValue(10)).toBe('Lower than 33');
      expect(rule.getFormattedValue(50)).toBe('Between 33 and 66');
      expect(rule.getFormattedValue(80)).toBe('Upper than 66');
    });
  });

  describe('Text replace', () => {
    const data = Rule.getDefaultData();
    const rule = new Rule('/.*/', data);
    const text = 'This is my value';
    const patternText = '/value/';
    const formattedValue = '12.34';
    const tm = rule.addTextMap('');
    test('All content', () => {
      expect(tm.getReplaceText(text, formattedValue)).toBe('12.34');
    });
    test('pattern', () => {
      tm.data.textReplace = 'pattern';
      tm.data.textPattern = patternText;
      expect(tm.getReplaceText('This is my value', formattedValue)).toBe('This is my 12.34');
    });
  });
});
