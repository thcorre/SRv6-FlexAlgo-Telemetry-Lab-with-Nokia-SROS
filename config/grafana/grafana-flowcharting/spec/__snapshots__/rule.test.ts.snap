// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rule New Should be equals 1`] = `
Object {
  "aggregation": "current",
  "alias": "myRule",
  "colors": Array [
    "rgba(245, 54, 54, 0.9)",
    "rgba(237, 129, 40, 0.89)",
    "rgba(50, 172, 45, 0.97)",
  ],
  "column": "Time",
  "dateFormat": "YYYY-MM-DD HH:mm:ss",
  "decimals": 2,
  "eventData": Array [],
  "eventProp": "id",
  "eventRegEx": false,
  "gradient": false,
  "hidden": false,
  "invert": false,
  "linkData": Array [],
  "linkProp": "id",
  "linkRegEx": true,
  "mappingType": 1,
  "metricType": "serie",
  "order": 1,
  "overlayIcon": false,
  "pattern": "/.*/",
  "rangeData": Array [],
  "reduce": true,
  "refId": "A",
  "sanitize": false,
  "shapeData": Array [
    Object {
      "colorOn": "a",
      "hidden": false,
      "pattern": "/.*Toto.*/",
      "style": "fillColor",
    },
  ],
  "shapeProp": "id",
  "shapeRegEx": true,
  "stringThresholds": Array [
    "/.*/",
    "/.*/",
  ],
  "textData": Array [],
  "textProp": "id",
  "textRegEx": true,
  "thresholds": Array [
    50,
    80,
  ],
  "tooltip": false,
  "tooltipColors": false,
  "tooltipLabel": "",
  "tooltipOn": "a",
  "tpDirection": "v",
  "tpGraph": false,
  "tpGraphHigh": null,
  "tpGraphLow": null,
  "tpGraphScale": "linear",
  "tpGraphSize": "100%",
  "tpGraphType": "line",
  "type": "number",
  "unit": "short",
  "valueData": Array [],
}
`;
